File: Merge Sort/merge_sort.rb
def merge_sort! (&prc)
  prc ||= Proc.new { |a, b| a <=> b }
  return self if self.length <= 1
  mid_index = self.length / 2
  left = self.take(mid_index).merge_sort!(&prc)
  rite = self.drop(mid_index).merge_sort!(&prc)
  return merge(left, rite, &prc)
end

def merge (left, rite, &prc)
    merged_array = []
    until left.empty? || rite.empty?
      case prc.call(left.first, rite.first)
      when  0; merged_array << left.shift
      when  1; merged_array << rite.shift
      when -1; merged_array << left.shift
      end
    end
    merged_array + left + rite
end

File: Merge Sort/MergeSort.cpp
#include "MergeSort.hpp"

// merge sort O(nlog(n))
std::vector<int> mergeSort(std::vector<int> array) {
  if (array.size() <= 1) return array;
  
  int midIndex = (int) array.size() / 2;
  std::vector<int> left;
  std::vector<int> rite;
  
  if (array.size() & 1) {
    left = std::vector<int>(array.begin(), array.end() - midIndex);
    rite = std::vector<int>(array.begin() + midIndex + 1, array.end());
  } else {
    left = std::vector<int>(array.begin(), array.end() - midIndex);
    rite = std::vector<int>(array.begin() + midIndex, array.end());
  }
  
  left = mergeSort(left);
  rite = mergeSort(rite);
  return merge(left, rite);
}

// merge helper function
std::vector<int> merge(std::vector<int> left, std::vector<int> rite) {
  std::vector<int> result;
  
  while (!left.empty() && !rite.empty()) {
      
    if (left[0] < rite[0]) {
      result.push_back(left[0]);
      left.erase(left.begin());
      
    } else if (left[0] > rite[0]) {
      result.push_back(rite[0]);
      rite.erase(rite.begin());
      
    } else {
      result.push_back(left[0]);
      left.erase(left.begin());
    }
  }
  
  result.insert(result.end(), left.begin(), left.end());
  result.insert(result.end(), rite.begin(), rite.end());
  return result;
}

// debugger
void print(std::vector<int> array) {
  std::cout << "{ ";
  for (int index = 0; index < array.size(); index++) {
    std::cout << array[index];
    if (index < array.size() - 1) std::cout << ", ";
  }
  std::cout << " }" << std::endl;
}
